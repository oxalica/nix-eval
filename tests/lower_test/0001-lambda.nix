[
  (x: x)
  (a: b: a + b)
  ({ a, b }: { c, d, ... }: a c + b d)
  ({ hello ? pkgs.hello, pkgs, world ? args, ... }@args:
    [ hello pkgs world ])
  (args@{ a ? 1, b }: a + args.b)
  (this:is:string)
]
