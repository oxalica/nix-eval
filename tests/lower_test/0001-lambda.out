Ok(
    Expr(
        List {
            items: [
                Expr(
                    Lambda {
                        arg: Bind,
                        body: Debruijn(
                            0,
                        ),
                    },
                ),
                Expr(
                    Lambda {
                        arg: Bind,
                        body: Expr(
                            Lambda {
                                arg: Bind,
                                body: Expr(
                                    BinOp {
                                        operator: Add,
                                        lhs: Debruijn(
                                            1,
                                        ),
                                        rhs: Debruijn(
                                            0,
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
                Expr(
                    Lambda {
                        arg: ClosePattern {
                            required_names: [
                                "a",
                                "b",
                            ],
                            optional_names: {},
                        },
                        body: Expr(
                            LetIn {
                                exprs: [
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                2,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            or_default: None,
                                        },
                                    ),
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                2,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            or_default: None,
                                        },
                                    ),
                                ],
                                body: Expr(
                                    Lambda {
                                        arg: OpenPattern {
                                            required_names: [
                                                "c",
                                                "d",
                                            ],
                                        },
                                        body: Expr(
                                            LetIn {
                                                exprs: [
                                                    Expr(
                                                        Select {
                                                            set: Debruijn(
                                                                2,
                                                            ),
                                                            index: Expr(
                                                                Literal(
                                                                    String(
                                                                        "c",
                                                                    ),
                                                                ),
                                                            ),
                                                            or_default: None,
                                                        },
                                                    ),
                                                    Expr(
                                                        Select {
                                                            set: Debruijn(
                                                                2,
                                                            ),
                                                            index: Expr(
                                                                Literal(
                                                                    String(
                                                                        "d",
                                                                    ),
                                                                ),
                                                            ),
                                                            or_default: None,
                                                        },
                                                    ),
                                                ],
                                                body: Expr(
                                                    BinOp {
                                                        operator: Add,
                                                        lhs: Expr(
                                                            Apply {
                                                                lambda: Debruijn(
                                                                    4,
                                                                ),
                                                                value: Debruijn(
                                                                    1,
                                                                ),
                                                            },
                                                        ),
                                                        rhs: Expr(
                                                            Apply {
                                                                lambda: Debruijn(
                                                                    3,
                                                                ),
                                                                value: Debruijn(
                                                                    0,
                                                                ),
                                                            },
                                                        ),
                                                    },
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
                Expr(
                    Lambda {
                        arg: OpenPattern {
                            required_names: [
                                "pkgs",
                            ],
                        },
                        body: Expr(
                            LetIn {
                                exprs: [
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                3,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "hello",
                                                    ),
                                                ),
                                            ),
                                            or_default: Some(
                                                Expr(
                                                    Select {
                                                        set: Debruijn(
                                                            1,
                                                        ),
                                                        index: Expr(
                                                            Literal(
                                                                String(
                                                                    "hello",
                                                                ),
                                                            ),
                                                        ),
                                                        or_default: None,
                                                    },
                                                ),
                                            ),
                                        },
                                    ),
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                3,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "pkgs",
                                                    ),
                                                ),
                                            ),
                                            or_default: None,
                                        },
                                    ),
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                3,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "world",
                                                    ),
                                                ),
                                            ),
                                            or_default: Some(
                                                Debruijn(
                                                    3,
                                                ),
                                            ),
                                        },
                                    ),
                                ],
                                body: Expr(
                                    List {
                                        items: [
                                            Debruijn(
                                                2,
                                            ),
                                            Debruijn(
                                                1,
                                            ),
                                            Debruijn(
                                                0,
                                            ),
                                        ],
                                    },
                                ),
                            },
                        ),
                    },
                ),
                Expr(
                    Lambda {
                        arg: ClosePattern {
                            required_names: [
                                "b",
                            ],
                            optional_names: {
                                "a",
                            },
                        },
                        body: Expr(
                            LetIn {
                                exprs: [
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                2,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            or_default: Some(
                                                Expr(
                                                    Literal(
                                                        Integer(
                                                            1,
                                                        ),
                                                    ),
                                                ),
                                            ),
                                        },
                                    ),
                                    Expr(
                                        Select {
                                            set: Debruijn(
                                                2,
                                            ),
                                            index: Expr(
                                                Literal(
                                                    String(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            or_default: None,
                                        },
                                    ),
                                ],
                                body: Expr(
                                    BinOp {
                                        operator: Add,
                                        lhs: Debruijn(
                                            1,
                                        ),
                                        rhs: Expr(
                                            Select {
                                                set: Debruijn(
                                                    2,
                                                ),
                                                index: Expr(
                                                    Literal(
                                                        String(
                                                            "b",
                                                        ),
                                                    ),
                                                ),
                                                or_default: None,
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
                Expr(
                    Literal(
                        String(
                            "this:is:string",
                        ),
                    ),
                ),
            ],
        },
    ),
)